File: app/src/main/java/com/kickstarter/ui/viewholders/EmptyCommentsViewHolder.java
package com.kickstarter.ui.viewholders;

import android.util.Pair;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.kickstarter.R;
import com.kickstarter.models.Project;
import com.kickstarter.models.User;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.kickstarter.libs.utils.ObjectUtils.requireNonNull;

public final class EmptyCommentsViewHolder extends KSViewHolder {
  private Project project;
  private User user;
  private final Delegate delegate;
  protected @Bind(R.id.comments_login_button) Button commentsLoginButton;
  protected @Bind(R.id.no_comments_message) TextView noCommentsMessageTextView;

  public interface Delegate {
    void emptyCommentsLoginClicked(EmptyCommentsViewHolder viewHolder);
  }

  public EmptyCommentsViewHolder(final @NonNull View view, final @NonNull Delegate delegate) {
    super(view);
    this.delegate = delegate;
    ButterKnife.bind(this, view);
  }

  @Override
  public void bindData(final @Nullable Object data) throws Exception {
    @SuppressWarnings("unchecked")
    final Pair<Project, User> projectAndUser = requireNonNull((Pair<Project, User>) data);
    this.project = requireNonNull(projectAndUser.first, Project.class);
    this.user = projectAndUser.second;
  }

  public void onBind() {
    if (this.user == null) {
      this.commentsLoginButton.setVisibility(View.VISIBLE);
      this.noCommentsMessageTextView.setText(R.string.project_comments_empty_state_logged_out_message_log_in);
    } else if (this.project.isBacking()) {
      this.commentsLoginButton.setVisibility(View.GONE);
      this.noCommentsMessageTextView.setText(R.string.project_comments_empty_state_backer_message);
    } else {
      this.commentsLoginButton.setVisibility(View.GONE);
      this.noCommentsMessageTextView.setText(R.string.update_comments_empty_state_non_backer_message);
    }
  }

  @Nullable
  @OnClick(R.id.comments_login_button)
  public void emptyCommentsLogin() {
    this.delegate.emptyCommentsLoginClicked(this);
  }
}

File: app/src/main/java/com/kickstarter/ui/viewholders/EmptyCommentsViewHolder.kt
package com.kickstarter.ui.viewholders

import android.util.Pair
import android.view.View
import androidx.annotation.StringRes
import com.kickstarter.R
import com.kickstarter.databinding.EmptyCommentsLayoutBinding
import com.kickstarter.libs.utils.ObjectUtils
import com.kickstarter.models.Project
import com.kickstarter.models.User

class EmptyCommentsViewHolder(private val binding: EmptyCommentsLayoutBinding, private val delegate: Delegate) : KSViewHolder(binding.root) {
    private var project: Project? = null
    private var user: User? = null

    interface Delegate {
        fun emptyCommentsLoginClicked(viewHolder: EmptyCommentsViewHolder?)
    }

    @Throws(Exception::class)
    override fun bindData(data: Any?) {
        val projectAndUser = ObjectUtils.requireNonNull(data as? Pair<Project, User>?)
        project = ObjectUtils.requireNonNull(projectAndUser.first, Project::class.java)
        user = projectAndUser.second
    }

    override fun onBind() {
        when {
            user == null -> {
                bindView(View.VISIBLE, R.string.project_comments_empty_state_logged_out_message_log_in)
            }
            project?.isBacking == true -> {
                bindView(View.GONE, R.string.project_comments_empty_state_backer_message)
            }
            else -> {
                bindView(View.GONE, R.string.update_comments_empty_state_non_backer_message)
            }
        }
        binding.commentsLoginButton.setOnClickListener {
            emptyCommentsLogin()
        }
    }
    private fun bindView(hasVisibility: Int, @StringRes string: Int) {
        binding.commentsLoginButton.visibility = hasVisibility
        binding.noCommentsMessage.setText(string)
    }

    fun emptyCommentsLogin() {
        delegate.emptyCommentsLoginClicked(this)
    }
}
